drivers/adc/adc_ch32v00x.c

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (c) 2025 Michael Hope
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT wch_adc
+
+#include <hal_ch32fun.h>
+
+#include <zephyr/drivers/adc.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/clock_control.h>
+
+struct adc_ch32v00x_config {
+	ADC_TypeDef *regs;
+	const struct pinctrl_dev_config *pin_cfg;
+	const struct device *clock_dev;
+	uint8_t clock_id;
+};
+
+static int adc_ch32v00x_channel_setup(const struct device *dev,
+				      const struct adc_channel_cfg *channel_cfg)
+{
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		return -EINVAL;
+	}
+	if (channel_cfg->reference != ADC_REF_INTERNAL) {
+		return -EINVAL;
+	}
+	if (channel_cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		return -EINVAL;
+	}
+	if (channel_cfg->differential) {
+		return -EINVAL;
+	}
+	if (channel_cfg->channel_id >= 10) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int adc_ch32v00x_read(const struct device *dev, const struct adc_sequence *sequence)
+{
+	const struct adc_ch32v00x_config *config = dev->config;
+	ADC_TypeDef *regs = config->regs;
+	uint32_t channels = sequence->channels;
+	int rsqr = 2;
+	int sequence_id = 0;
+	int total_channels = 0;
+	int i;
+	uint16_t *samples = sequence->buffer;
+
+	if (sequence->options != NULL) {
+		return -ENOTSUP;
+	}
+	if (sequence->resolution != 10) {
+		return -EINVAL;
+	}
+	if (sequence->oversampling != 0) {
+		return -ENOTSUP;
+	}
+	if (sequence->channels >= (1 << 10)) {
+		return -EINVAL;
+	}
+
+	if (sequence->calibrate) {
+		regs->CTLR2 |= ADC_RSTCAL;
+		while ((regs->CTLR2 & ADC_RSTCAL) != 0) {
+		}
+		regs->CTLR2 |= ADC_CAL;
+		while ((regs->CTLR2 & ADC_CAL) != 0) {
+		}
+	}
+
+	/*
+	 * Build the sample sequence. The channel IDs are packed 5 bits at a time starting in RSQR3
+	 * and working down in memory to RSQR1.
+	 */
+	regs->RSQR1 = 0;
+	regs->RSQR2 = 0;
+	regs->RSQR3 = 0;
+
+	for (i = 0; channels != 0; i++, channels >>= 1) {
+		if ((channels & 1) != 0) {
+			total_channels++;
+			(&regs->RSQR1)[rsqr] |= i << sequence_id;
+			sequence_id += ADC_SQ2_0;
+			if (sequence_id >= 32) {
+				/* Move on to the next RSQRn register, i.e. RSQR(n-1) */
+				sequence_id = 0;
+				rsqr--;
+			}
+		}
+	}
```

Comment:
i might be wrong but I am not sure it's handling the "boundaries" properly? ie. for SQ7 you are writting starting from reserved bit 30 or RSQR3?

Reply:
Fixed. This should be at most 6 channels at 5 bits/channel. Changed the ADC_SQ2_0 for the explicit number of bits.